### PR TITLE
[lldb] Make conversions from llvm::Error explicit with Status::FromEr…

### PR DESCRIPTION
**Pull request by adrian-prantl**
…ror() [NFC]

Broken out from https://github.com/llvm/llvm-project/pull/106774 for easier review.

### CODE REVIEW

**Review by labath — COMMENTED** [review]
I don't think this usage of rvalue references is very idiomatic. It saves an object copy (move), but:
- that should be very cheap anyway
- it increases the risk of improper handling and weird side-effects -- a function like `void use(Error E) {}` will always crash at the closing bracket (due to not checking the error), but `use(Error &&E){}` will never crash *inside* the function (because it doesn't own the error), and whether the program as a whole crashes depends on whether the caller does any other operation that would clear the error object.

So, I think it would be better to stick to value semantics. Other than that, this seems fine.

**Comment by adrian-prantl** [comment]
> I don't think this usage of rvalue references is very idiomatic. It saves an object copy (move), but:
> 
> * that should be very cheap anyway
> * it increases the risk of improper handling and weird side-effects -- a function like `void use(Error E) {}` will always crash at the closing bracket (due to not checking the error), but `use(Error &&E){}` will never crash _inside_ the function (because it doesn't own the error), and whether the program as a whole crashes depends on whether the caller does any other operation that would clear the error object.
> 
> So, I think it would be better to stick to value semantics. Other than that, this seems fine.

@labath Are you suggesting to replace all assignments of Status (everything in this patch) with calls to Clone()? In https://github.com/llvm/llvm-project/pull/106774 I'll change Status to wrap llvm::Error(), which can *only* be moved. While the end-goal is to replace most uses of Status with llvm::Error, we're going to be stuck with a lot of APIs that use Status during the transition and that would be a lot of unnecessary copying. And in all APIs we convert to llvm::Error, we'll also have to use the move semantics for assignment.

File: lldb/include/lldb/Utility/Status.h [hunk 1]
```diff
@@ -91,9 +91,9 @@ class Status {
 
   ~Status();
 
-  // llvm::Error support
-  explicit Status(llvm::Error error) { *this = std::move(error); }
-  const Status &operator=(llvm::Error error);
+  /// Avoid using this in new code. Migrate APIs to llvm::Expected instead.
+  static Status FromError(llvm::Error &&error);
```
Comment (labath) [inline lldb/include/lldb/Utility/Status.h:95]:
```suggestion
  static Status FromError(llvm::Error error);
```

File: lldb/include/lldb/Utility/Status.h [hunk 2]
```diff
@@ -145,6 +145,7 @@ class Status {
   bool Success() const;
 
 protected:
+  Status(llvm::Error &&error);
```
Comment (labath) [inline lldb/include/lldb/Utility/Status.h:148]:
```suggestion
  Status(llvm::Error error);
```

**Review by labath — COMMENTED** [review]
No, I'm just saying we should ditch the rvalue references. I.e., do this.

**Comment by adrian-prantl** [comment]
@labath Done.

**Review by labath — APPROVED** [review]
Thanks. Looks good. There are a bunch of linux failures reported by the linux precommit bot. I don't expect you'll get all of them, but you could try to get at least those that are reported there so that there's one less post-commit iteration.

**Comment by nico** [comment]
Looks like this breaks building on windows: http://45.33.8.238/win/93597/step_3.txt

Please take a look and revert for now if it takes a while to fix.

**Pull request merged**